### PR TITLE
Add config to exempt local vars from ExplicitTypeInterfaceRule

### LINKED ISSUE #2028: [Swift 4.1] Add configuration option to fire `explicit_type_interface` only for properties rather than local variables
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.24.2
* Installation method used (Homebrew, CocoaPods, building from source, etc)? portable
* Paste your configuration file:

* Which Xcode version are you using (check `xcode-select -p`)? 9.3

```swift
// This triggers a violation:
func a() {
    let foo = UIView()
}
```

With reference to issue #2027

### PR DESCRIPTION
**Pull request by rounak**
This PR adds a new `excluded` config to this existing rule, to exempt certain kinds of variables from having explicit types. Values can be `local`, `static`, `class` & `instance`.

Would love to add test coverage for this, but didn't know if I should write a new test in `RuleTests` or do it some other way.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,7 +6,10 @@
 
 #### Enhancements
 
-* None.
+* Add a new `excluded` config parameter to the `explicit_type_interface` rule
+  to exempt certain types of variables from the rule.
```
Comment (marcelofabri) [inline CHANGELOG.md]: you should leave two trailing whitespaces (currently just one)

File: Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        default:
+            return nil
+        }
+    }
+}
+
+public struct ExplicitTypeInterfaceConfiguration: RuleConfiguration, Equatable {
+    public var severityConfiguration = SeverityConfiguration(.warning)
+
+    var allowedKinds: Set<SwiftDeclarationKind> = [.varInstance,
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift]: this should be `public`

File: Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                                                   .varLocal,
+                                                   .varStatic,
+                                                   .varClass]
+
+    public var consoleDescription: String {
+        return severityConfiguration.consoleDescription + ", allowed kinds: \(allowedKinds)"
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift]: can we print the allowedKinds in the same short style that is expect to be in the configuration file?

**Comment by marcelofabri** [comment]
> Would love to add test coverage for this, but didn't know if I should write a new test in RuleTests or do it some other way.

You can create a new file `ExplicitTypeInterfaceConfigurationTests` (take a look at `CyclomaticComplexityConfigurationTests` for example) and another one to test the rule itself (`ExplicitTypeInterfaceRuleTests`).

When you create `ExplicitTypeInterfaceRuleTests`, you should move the existing test in `RulesTests` to the new file, along with new tests that test different configurations. You can take a look at existing tests to see how it's usually done (`ColonRuleTests` would be an example).

File: Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift [hunk 4]
```diff
@@ -27,27 +27,37 @@ private extension SwiftDeclarationKind {
...
-    var allowedKinds: Set<SwiftDeclarationKind> = [.varInstance,
-                                                   .varLocal,
-                                                   .varStatic,
-                                                   .varClass]
+    public var allowedKinds = ExplicitTypeInterfaceConfiguration.variableKinds
 
     public var consoleDescription: String {
-        return severityConfiguration.consoleDescription + ", allowed kinds: \(allowedKinds)"
+        return severityConfiguration.consoleDescription +
+        ", excluded: [\(ExplicitTypeInterfaceConfiguration.variableKinds.subtracting(allowedKinds))]"
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift]: My comment here was more about how the types will be spelled out.

By just printing `SwiftDeclarationKind`, you'll get something like `source.lang.swift.decl.var.class`, but when configuring the rule you're expected to use just `class`.

**Comment by rounak** [comment]
Thanks for the review @marcelofabri.

Added two new test files as you said, one for config and one for the rule itself. The only issue is since this local variable stuff depends on SourceKit, we can't add concrete test cases (it warns for local vars) until everyone's running Xcode 9.3.

As for the consoleDescription, I didn't quite understand:

> the same short style that is expect to be in the configuration file

I added square brackets around the set, is that what you meant?

File: Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift (hunk 4 above)
Comment (rounak) [reply Source/SwiftLintFramework/Rules/RuleConfigurations/ExplicitTypeInterfaceConfiguration.swift]: Ah I see, let me do that

**Comment by marcelofabri** [comment]
> The only issue is since this local variable stuff depends on SourceKit, we can't add concrete test cases (it warns for local vars) until everyone's running Xcode 9.3.

Can you add a test excluding another kind (e.g. class variables)?

**Comment by rounak** [comment]
@marcelofabri Done, let me know if it looks okay now.

File: Tests/SwiftLintFrameworkTests/ExplicitTypeInterfaceTests.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+//  SwiftLint
+//
+//  Created by Rounak Jain on 2/24/18.
+//  Copyright © 2018 Realm. All rights reserved.
+//
+
+import SwiftLintFramework
+import XCTest
+
+class ExplicitTypeInterfaceTests: XCTestCase {
```
Comment (marcelofabri) [inline Tests/SwiftLintFrameworkTests/ExplicitTypeInterfaceTests.swift]: Can you rename this class (and file) to `ExplicitTypeInterfaceRuleTests`?

**Comment by marcelofabri** [comment]
@rounak This looks good! I have one minor nitpick that I added as a comment.

Can you also rebase against master now that I merged #2068? That should fix the CI failure.

Thanks!

**Pull request merged**